development/cpp/internal_rendering_architecture.rst

```diff
@@ -0,0 +1,600 @@
+
+Forward rendering generally provides a better tradeoff for performance versus
+flexibility, especially when a clustered approach to lighting is used. While
+deferred rendering can be faster in some cases, it's also less flexible and
+requires using hacks to be able to use MSAA. Since games with a less realistic
+art style can benefit a lot from MSAA, we chose to go with forward rendering for
+Godot 4 (like Godot 3).
+
+That said, parts of the forward renderer *are* performed with a deferred approach
+to allow for some optimizations when possible. This applies to VoxelGI and SDFGI
+in particular.
+
+A clustered deferred renderer may be developed in the far future. This renderer
+could be used in situations where performance is favored over flexibility.
+
+Summary of rendering drivers/backends
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The following rendering API + backend combinations are currently possible:
+
+- Vulkan + Clustered Forward
+- Vulkan + Forward Mobile
+- Direct3D 12 + Clustered Forward
+- Direct3D 12 + Forward Mobile
+- Metal + Clustered Forward (via MoltenVK)
+- Metal + Forward Mobile (via MoltenVK)
+- OpenGL + Compatibility
+
+Each combination has its own limitations and performance characteristics. Make
+sure to test your changes on all backends if possible before opening a pull
+request.
+
+RenderingDevice abstraction
+---------------------------
+
+.. note::
+
+    The OpenGL driver does not use the RenderingDevice abstraction.
+
+To make the complexity of modern low-level graphics APIs more manageable,
+Godot uses its own abstraction called RenderingDevice.
+
+This means that when writing code for modern rendering backends, you don't
+actually use the Vulkan or Direct3D 12 APIs directly. While this is still
+lower-level than an API like OpenGL, this makes working on the renderer easier,
+as RenderingDevice will abstract many API-specific quirks for you.
+
+**Vulkan RenderingDevice implementation:**
+
+- `drivers/vulkan/rendering_device_vulkan.cpp <https://github.com/godotengine/godot/blob/master/drivers/vulkan/rendering_device_vulkan.cpp>`__
+
+.. _doc_internal_rendering_architecture_core_shaders:
+
+Core shaders
+------------
+
+While shaders in Godot projects are written using a
+:ref:`custom language inspired by GLSL <doc_shading_language>`, core shaders are
+written directly in GLSL.
+
+These core shaders are embedded in the editor and export template binaries at
+compile-time. To see any changes you've made to those GLSL shaders, you need to
+recompile the editor or export template binary.
+
+Some material features such as height mapping, refraction and proximity fade are
+not part of core shaders, and are performed in the default BaseMaterial3D using
+the Godot shader language instead (not GLSL). This is done by procedurally
+generating the required shader code depending on the features enabled in the
+material.
+
+By convention, shader files with ``_inc`` in their name are included in other
```

Comment:
Maybe also mention that adding include files or more features into the core shaders should only be done if absolutely necessary since they need to compile very quickly? At least that's what I understood from Juan's feedback on my Visual Shader PR: https://github.com/godotengine/godot/pull/64248#issuecomment-1214167428

Reply:
Done :slightly_smiling_face: